core/shared/src/main/scala/zio/ZIOFunctionConstructor.scala

```diff
@@ -0,0 +1,104 @@
+package zio
+
+/**
+ * A `ZIOFunctionConstructor[Input]` knows how to construct a `ZIO` value from a
+ * function of type `In`. This allows the type of the `ZIO` value constructed to
+ * depend on `In`.
+ */
+sealed abstract class ZIOFunctionConstructor[In] {
+
+  /**
+   * The type of the `ZIO` value.
+   */
+  type Out
+
+  /**
+   * Constructs a `ZIO` value from the specified input.
+   */
+  def apply(in: In)(implicit trace: Trace): Out
+}
+
+object ZIOFunctionConstructor {
+
+  type WithOut[In, Out0] = ZIOFunctionConstructor[In] { type Out = Out0 }
+
+  implicit def function1Constructor[A: Tag, Z: Tag, R, E]
```

Comment:
Really sorry for not picking this earlier and causing extra rounds of reviews/comments.

I think the `Tag` implicit on Z is unnecessary (same for all other methods)

```suggestion
  implicit def function1Constructor[A: Tag, Z, R, E]
```

Reply:
Also by convention we generally place the output type param at the rightmost side, i.e., `[A: Tag, R, E, Z]`

Reply:
done